updated tree-sitter grammar to support bpf map field macros

ivzap opened the pull request:
bpf map field macros at the current projects state are unsupported. To add this a change in tree-sitter grammar was required. That is, a new construct that supports the usage of bpf macros was added into grammar.js. This was achieved by creating a new tree-sitter rule called `bpf_map_field_macro` and added as a choice in the `_field_declaration_list_item` rule. In other words, when tree-sitter sees a macro of the form __macro(a,b,...) it will create a new node within the tree called `bpf_map_field_macro`. We can now write a query that checks for `bpf_map_field_macro` nodes and the usage of sizeof() in the second argument to provide the lint.

Additionally, a couple tests were added to /test/corpus/bpf.txt

d-e-s-o commented on grammar.js:
```diff
@@ -28,5 +29,43 @@ module.exports = grammar(C, {
+      seq(
+        token(prec(1, /__\w+/)),
+        '(',
+        optional(commaSep($.identifier)),
```
Hm, this seems bogus: Why is there an `optional` and then inside `commaSep` another one? Given that we only use the `commaSep` function once, perhaps it's not worth having a function and we can just inline it here instead?

d-e-s-o commented on grammar.js:
```diff
@@ -28,5 +29,43 @@ module.exports = grammar(C, {
+
+    bpf_map_field_macro: $ => choice(
+      seq(
+        token(prec(1, /__\w+/)),
```
Why do we need this `prec` here? It's not clear to me.

In general, aren't we looking for a macro invocation, so I would kind of expect the usage of `preproc_call` here. Have you tried that?

Even if that doesn't work out, I think that we shouldn't be defining a custom token here. It's a macro call, and macros are defined as something like
```
    preproc_def: $ => seq(
      preprocessor('define'),
      field('name', $.identifier),
      field('value', optional($.preproc_arg)),
      token.immediate(/\r?\n/),
    ),
```
So aren't we just looking for an `identifier`?

d-e-s-o commented on test/corpus/bpf.txt:
```diff
@@ -71,3 +71,95 @@ int BPF_KPROBE(kprobe__cap_capable, const struct cred *cred,
             (identifier)))))))
+
+==================
+uint macro bad usage
```
It's just a grammar. There isn't really bad usage if the code compiles, I'd say.
```suggestion
__uint macro with sizeof operator
```

d-e-s-o commented on test/corpus/bpf.txt:
```diff
@@ -71,3 +71,95 @@ int BPF_KPROBE(kprobe__cap_capable, const struct cred *cred,
+              (primitive_type)
+              (pointer_declarator
+                (field_identifier)))))
+        (identifier)))
```
Fair enough, but I am 99% sure that this case is covered by the existing C grammar tests, so I'd think we can drop it. E.g., see `Struct declarations` test.

d-e-s-o commented on test/corpus/bpf.txt:
```diff
@@ -71,3 +71,95 @@ int BPF_KPROBE(kprobe__cap_capable, const struct cred *cred,
+        (identifier)))
+
+==================
+type macro usage
```
```suggestion
__type macro usage
```

d-e-s-o commented on test/corpus/bpf.txt:
```diff
@@ -71,3 +71,95 @@ int BPF_KPROBE(kprobe__cap_capable, const struct cred *cred,
+            (field_declaration
+              (primitive_type)
+              (field_identifier))))
+        (identifier)))
```
Can we also have a test for different ordering of some field macro with an `int` attribute? Say the `int` attribute declaration is in the middle somewhere?

Please also add a test for more elaborate values, such as a complete expression (multiplication) as I pointed out in the existing `task_longrun.bpf.c` example.

d-e-s-o reviewed (changes requested):
Thanks for working on this feature! Can you please also include everything generated by `tree-sitter generate`? See for example https://github.com/d-e-s-o/tree-sitter-bpf-c/commit/02a115144348af1068e0af2d780fec335ab6c9e6

I think you'll have to generate it with `tree-sitter-cli` 0.25.8 for CI to pass (it's a bit annoying that they encode the minor version in there and that the workflow checks that, but it is what it is).

I also think the current `bpf_map_field_macro` definition is too constrained. For example, the following https://github.com/d-e-s-o/bpflint/blob/775fea29161a2237395c5e28e35fdec7aa9723f3/examples/task_longrun.bpf.c#L12 is not parsed correctly, it seems. I think the grammar should be flexible enough to capture this case. Can you check?

Commit:
updated tree-sitter grammar to support bpf map field macros

bpf map field macros at the current projects state are unsupported. To add this
a change in tree-sitter grammar was required. That is, the
_field_declaration_list_item rule now has an additional choice that checks for
a type_field_declarator_macro or a uint_field_declarator_macro which are
two new rules. With these rules a preproc_call_expression will be seen
within the tree to be parsed by a query.

d-e-s-o commented on test/corpus/bpf.txt:
```diff
@@ -71,3 +71,102 @@ int BPF_KPROBE(kprobe__cap_capable, const struct cred *cred,
+              (identifier)
+              (argument_list
+                (identifier)
+                (ERROR
```
Hm...we shouldn't be seeing an error here, but I think it's because something like `type ipv6hdr` isn't a valid construct inside a `call_expression`. We could probably live with that for the time being, but it would be better to do it right from the start. Do you think you can adjust the grammar further to accept that?

ivzap replied:
Yeah I could look into doing this. It looks like the grammar confuses when it sees a things of the form (type identifier)

ivzap replied:
Would it be ok if I created a new node for ```struct x``` arguments and if so what would you like it to be called?

d-e-s-o replied:
I'd think it would be about reusing an existing rule, right? `sizeof` for example should already accept such a construct as argument, so I'd check what it uses.

ivzap replied:
sizeof works well with handling that ```type ipv6hdr``` case since its within its own state if that makes sense. So a rule sizeof_expression was explicitly made for this function allowing it to not conflict with other states since the sizeof_expression state is more constrained/isolated. So I believe we would need to create a new rule for our macros which would allow us to control which argument constructs they accept at a more fine grain level. Checkout my changes and let me know what you think.

d-e-s-o commented on grammar.js:
```diff
@@ -28,5 +28,15 @@ module.exports = grammar(C, {
     ),
+    
+    _field_declaration_list_item: $ => choice(
+      seq($.call_expression, ';'),
```
I think strictly speaking the correct thing here would be `preproc_call_expression` instead. I think `call_expression` would probably allow constructs such as `(*foobar)()`, which makes no sense in that position.
But it seems as if it's more than a simple replace, as the latter seems to generate conflicts of sorts. If it's strictly accepting more that's probably fine, though I suspect there could be other issues (judging by the logic of the conflicts); not sure.

If you have an idea how to replace it then feel free to do so, but if not then let's leave it as-is and just add a TODO comment that this is likely not 100% correct.

ivzap replied:
Hmm, I feel like it would be fine if we allow the call_expression to be seen within the tree since if we can successfully parse the tree (I dont see any ERROR  in the tree during parsing that (*foobar)()) we can write a query to suggest lints based on however we want to conform to bpf standards? For example:

```c
struct{
	__uint((*foobar)(), b);
} name;
```
produces
```
(translation_unit
      (declaration
        (struct_specifier
          (field_declaration_list
            (call_expression
              (identifier)
              (argument_list
                (call_expression
                  (parenthesized_expression
                    (pointer_expression
                      (identifier)))
                  (argument_list))
                (identifier)))))
        (identifier)))
```

so in this case we could write a query that would produce a lint when that parenthesized expression is used if thats not allowed. It seems like both `preproc_call_expression` and `call_expression` share too many similar nodes that both could be produced so its ambiguous to the parser and only in some small cases can we truly tell if its a `preproc_call_expression`.  Lmk what you think?

d-e-s-o replied:
Perhaps I misunderstood, but my impression was that this node applies to the entire statement, `__uint((*foobar)(), b);`. I'd expect `preproc_call_expression` to accept `(*foobar)()` equally well in the position it is in. However, something like `(*foobar)(b);`would also have been accepted by `call_expression`, which I am not sure `preproc_call_expression` would. Anyway, it's different now.

d-e-s-o commented on test/corpus/bpf.txt:
```diff
@@ -71,3 +71,102 @@ int BPF_KPROBE(kprobe__cap_capable, const struct cred *cred,
+================================================================================
+
+struct{
+	__uint(type, BPF_MAP_TYPE_HASH);
```
Is the weird indentation here intentional or can it be fixed?

d-e-s-o commented on grammar.js:
```diff
@@ -28,5 +28,15 @@ module.exports = grammar(C, {
       field('declarator', $._declarator),
       field('body', $.compound_statement),
     ),
+    
```
Seems there is trailing spaces here, can you please remove them?

d-e-s-o reviewed (commented):
Looks mostly great now. One test contains an `ERROR`, though, when I think it shouldn't. Can you check if that can be fixed easily?

I think once that is addressed it would be good to get the full lint built up next before we merge this change, just so that we have confirmation that it actually all works together.

ivzap commented:
> I think once that is addressed it would be good to get the full lint built up next before we merge this change, just so that we have confirmation that it actually all works together.

Would you like me to add some query files to my bpflint fork, screenshots, not sure the best way to show its working?

d-e-s-o commented:
> > I think once that is addressed it would be good to get the full lint built up next before we merge this change, just so that we have confirmation that it actually all works together.
> 
> Would you like me to add some query files to my bpflint fork, screenshots, not sure the best way to show its working?

Can you just create a pull request for `bpflint`? It's fine if it fails in CI while it's in review. I can merge the `bpf-c` change then and publish a new release once we have a working lint.

d-e-s-o commented on grammar.js:
```diff
@@ -28,5 +28,41 @@ module.exports = grammar(C, {
+        '(',
+        $.identifier,
+        ',',
+        choice(prec(3, $.type_descriptor), prec(2, $.expression)), // TODO: $.expression is generic, choices needs to be more specific in future
```
Nit: Style-wise, let's please honor some line length limit of 100 characters (or 80, as you seemingly did below) or so, max, and keep comments on the previous line.

```suggestion
        // TODO: $.expression is generic, choices needs to be more specific in future
        choice(prec(3, $.type_descriptor), prec(2, $.expression)),
```

Personally, I'd remove the comment altogether. I fail to see the issue here; we don't have to be overly prescriptive about what is being accepted. It seems that if anything `expression` is more flexible than need be (though I am not convinced it is?), which I'd think is fine. Or am I missing something?

That being said, still on this line, why do we need these `prec` invocations? Everything seems fine without.

ivzap replied:
I'll remove the comment, I misunderstood your original comment about (*foo)() types being an invalid argument and was commenting `$.expression` would allow that. I'll remove prec here as I see your point, I did this to preemptively avoid conflicts but I see the parse doesn't warn of conflicts after I remove.

d-e-s-o commented on grammar.js:
```diff
@@ -28,5 +28,41 @@ module.exports = grammar(C, {
+        '(',
+        $.identifier,
+        ',',
+        choice(prec(3, $.type_descriptor), prec(2, $.type_specifier), prec(1, $.identifier)),
```
Same question as above about `prec` everywhere.

d-e-s-o reviewed (approved):
Changes look good to me, thanks! Can you squash changes into a single commit again? Also, I still see trailing white spaces in two places. I'd recommend adjusting your editor (or at least Git) to at least highlight those. Lastly, please also line-wrap commit messages at 80 characters or so, as is common practice.

ivzap commented:
> Changes look good to me, thanks! Can you squash changes into a single commit again? Also, I still see trailing white spaces in two places. I'd recommend adjusting your editor (or at least Git) to at least highlight those. Lastly, please also line-wrap commit messages at 80 characters or so, as is common practice. 

I'll make these changes to my editor thank you!

d-e-s-o commented:
Can you please check why the newly added tests fail?

ivzap commented:
> Can you please check why the newly added tests fail?

Yeah thats strange. I believe its because my corpus/bpf.txt is an old version, not sure how that happend. I'll make the change.

d-e-s-o commented:
Merged with a few modifications as https://github.com/d-e-s-o/tree-sitter-bpf-c/commit/96c04bda7880a4930c07b29fe9cd47b3edc43c82
